From: Eduard Bloch <blade@debian.org>
Date: Sun, 08 May 2022 23:03:35 +0200
Subject: Linux 5.17 support
Origin: https://raw.githubusercontent.com/archlinux/svntogit-community/master/broadcom-wl-dkms/trunk/012-linux517.patch

Suggested in https://bugs.debian.org/1010128, its original commit information:

6.30.223.271-29: add patch for linux 5.17
git-svn-id: file:///srv/repos/svn-community/svn@1171161 9fca08f4-af9d-4005-b8df-a31f2cc04f65
heftig committed on 23 Mar 

diff -u -r a/amd64/src/wl/sys/wl_linux.c b/amd64/src/wl/sys/wl_linux.c
--- a/amd64/src/wl/sys/wl_linux.c	2022-03-23 00:35:42.930416350 +0000
+++ b/amd64/src/wl/sys/wl_linux.c	2022-03-23 00:40:12.903771013 +0000
@@ -2980,7 +2980,11 @@
 	else
 		dev->type = ARPHRD_IEEE80211_RADIOTAP;
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 17, 0)
 	bcopy(wl->dev->dev_addr, dev->dev_addr, ETHER_ADDR_LEN);
+#else
+	eth_hw_addr_set(wl->dev, dev->dev_addr);
+#endif
 
 #if defined(WL_USE_NETDEV_OPS)
 	dev->netdev_ops = &wl_netdev_monitor_ops;
@@ -3261,7 +3265,11 @@
 static ssize_t
 wl_proc_read(struct file *filp, char __user *buffer, size_t length, loff_t *offp)
 {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 17, 0)
 	wl_info_t * wl = PDE_DATA(file_inode(filp));
+#else
+	wl_info_t * wl = pde_data(file_inode(filp));
+#endif
 #endif
 	int bcmerror, len;
 	int to_user = 0;
@@ -3318,7 +3326,11 @@
 static ssize_t
 wl_proc_write(struct file *filp, const char __user *buff, size_t length, loff_t *offp)
 {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 17, 0)
 	wl_info_t * wl = PDE_DATA(file_inode(filp));
+#else
+	wl_info_t * wl = pde_data(file_inode(filp));
+#endif
 #endif
 	int from_user = 0;
 	int bcmerror;
